From e9c33f08be0c6961a4e09b116281805e31eff29b Mon Sep 17 00:00:00 2001
From: Pino Toscano <ptoscano@redhat.com>
Date: Mon, 30 Jun 2025 14:10:16 +0200
Subject: [PATCH 1/3] tests: port pytest_ignore_collect() to pathlib

Create a pathlib version of the "path" argument of the
pytest_ignore_collect() hook; this will make it easier to the newer
version that uses pathlib directly.

This makes it possible to simplify the check for filenames to ignore:
since the filename is available, use it to do a quick lookup.

Signed-off-by: Pino Toscano <ptoscano@redhat.com>
---
 tests/conftest.py | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/tests/conftest.py b/tests/conftest.py
index 7ec7adf100..38cb09f1f5 100644
--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -2,6 +2,7 @@
 # See the COPYING file in the top-level directory.
 
 import os
+import pathlib
 
 import pytest
 
@@ -53,6 +54,7 @@ def pytest_addoption(parser):
 
 
 def pytest_ignore_collect(path, config):
+    collection_path = pathlib.Path(path)
     uitests_requested = config.getoption("--uitests")
 
     # Default --uitests to --verbosity=2
@@ -60,14 +62,14 @@ def pytest_ignore_collect(path, config):
         config.option.verbose = max(2, config.option.verbose)
 
     # Unless explicitly requested, ignore these tests
-    if "test_dist.py" in str(path):
-        return True
-    if "test_urls.py" in str(path):
-        return True
-    if "test_inject.py" in str(path):
+    if collection_path.name in (
+        "test_dist.py",
+        "test_urls.py",
+        "test_inject.py",
+    ):
         return True
 
-    uitest_file = "tests/uitests" in str(path)
+    uitest_file = "tests/uitests" in str(collection_path)
     if uitest_file and not uitests_requested:
         return True
     if not uitest_file and uitests_requested:

From 9754da8f0dce4b93e007cb1f618302c2a3e12428 Mon Sep 17 00:00:00 2001
From: Pino Toscano <ptoscano@redhat.com>
Date: Mon, 30 Jun 2025 14:20:36 +0200
Subject: [PATCH 2/3] tests: handle "collection_path" and "path" for
 pytest_ignore_collect()

Starting from pytest 7, all the hooks that take a "path" (the legacy
path data type) as parameter (including pytest_ignore_collect()) now
take a "collection_path" (pathlib.Path), and the latter will be the
only option starting from pytest 9.

Since it looks like pluggy (the plugin/hook infrastructure used by
pytest underneath) does not support using "None" to accept extra
arguments in an hook, then the solution is to provide different hooks
according to the pytest version:
- the common implementation is switched to use pathlib.Path, looking
  like the new style of hook; once pytest 7 is assumed, this helper
  function can be switched back to be the actual hook
- declare pytest_ignore_collect() as supported by pytest, using
  "collection_path" as earlier as possible; pytest_ignore_collect()
  is either a no-op wrapper, or builds a pathlib.Path out of the legacy
  type
- pytest.version_tuple was added in pytest 7, so deal with its lack in
  older versions

All in all, there should be no behaviour change in the actual test
collection, and the deprecation warnings should be gone.

Signed-off-by: Pino Toscano <ptoscano@redhat.com>
---
 tests/conftest.py | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/tests/conftest.py b/tests/conftest.py
index 38cb09f1f5..3c07311381 100644
--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -53,8 +53,7 @@ def pytest_addoption(parser):
     )
 
 
-def pytest_ignore_collect(path, config):
-    collection_path = pathlib.Path(path)
+def _impl_pytest_ignore_collect(collection_path, config):
     uitests_requested = config.getoption("--uitests")
 
     # Default --uitests to --verbosity=2
@@ -76,6 +75,17 @@ def pytest_ignore_collect(path, config):
         return True
 
 
+if getattr(pytest, "version_tuple", (0,)) >= (7,):
+
+    def pytest_ignore_collect(collection_path, config):
+        return _impl_pytest_ignore_collect(collection_path, config)
+
+else:
+
+    def pytest_ignore_collect(path, config):
+        return _impl_pytest_ignore_collect(pathlib.Path(path), config)
+
+
 def pytest_collection_modifyitems(config, items):
     def find_items(basename):
         return [i for i in items if os.path.basename(i.fspath) == basename]

From 932ec4f50ef627bc7942e69e695cf1f117aa2c65 Mon Sep 17 00:00:00 2001
From: Pino Toscano <ptoscano@redhat.com>
Date: Mon, 30 Jun 2025 14:56:08 +0200
Subject: [PATCH 3/3] tests: use Node.path when available

pytest 7 adds a "path" attribute to the Node class, which is the base of
the Item class used for items passed to pytest_collection_modifyitems().

Hence use it when available, using an helper function that tries to use
it and fallbacks on the old "fspath".

Signed-off-by: Pino Toscano <ptoscano@redhat.com>
---
 tests/conftest.py | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/tests/conftest.py b/tests/conftest.py
index 3c07311381..88a710fdd8 100644
--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -87,8 +87,14 @@ def pytest_ignore_collect(path, config):
 
 
 def pytest_collection_modifyitems(config, items):
+    def item_path_basename(item):
+        try:
+            return item.path.name
+        except AttributeError:
+            return os.path.basename(item.fspath)
+
     def find_items(basename):
-        return [i for i in items if os.path.basename(i.fspath) == basename]
+        return [i for i in items if item_path_basename(i) == basename]
 
     # Move test_cli cases to the end, because they are slow
     # Move test_checkprops to the very end, because it needs to run
